chipflow_lib/platforms/utils.py

```diff
@@ -0,0 +1,310 @@
+        n: Length of consecutive sequence to find
+        
+    Returns:
+        A slice indexing the first sequence of n consecutive numbers found within the given list
+        if unable to find a consecutive list, allocate as contigously as possible
+    """
+    if not lst or len(lst) < total:
+        raise ChipFlowError("Invalid request to find_contiguous_argument")
+
+    grouped = _group_consecutive_items(lst)
+
+    ret = []
+    n = total
+    for k in sorted(grouped.keys(), reverse=True):
+        for g in grouped[k]:
+            assert (n+len(ret)==total)
+            if k >= n:
+                ret += g[0:min(n,k)]
+                return ret
+            else:
+                n = n - k
+                ret += g[0:k]
+
+    return ret
+
+
+class _BasePackageDef(pydantic.BaseModel, abc.ABC):
+    # Used by pydantic to differentate when deserialising,
+    # override appropriately when you subclass
+    type: Literal["_BasePackageDef"] = "_BasePackageDef"
+    name: str
+
+    @property
+    @abc.abstractmethod
+    def pins(self) -> Set[str]:
+        ...
+    
+    @abc.abstractmethod
+    def allocate(self, available: Set[str], width:int) -> List[str]:
+        ...
+
+    @abc.abstractmethod
+    def sortpins(self, pins: Union[List[str], Set[str]]):
+        ...
+
+
+class _PGAPackageDef(_BasePackageDef):
+    # Used by pydantic to differentate when deserialising
+    type: Literal["_PGAPackageDef"] = "_PGAPackageDef"
+
+    size: int
+
+    @property
+    def pins(self) -> Set[str]:
+        return set([str(i) for i in range(self.size-1)])
+
+    def allocate(self, available: Set[str], width:int) -> List[str]:
+        avail_n = sorted([int(i) for i in available])
+        logger.debug(f"PGAPackageDef.allocate {width} from {len(avail_n)} remaining")
+        ret = _find_contiguous_sequence(avail_n, width)
+        logger.debug(f"PGAPackageDef.returned {ret}")
+        assert(len(ret)==width)
+        return [str(i) for i in ret]
+        
+    def sortpins(self, pins: Union[List[str], Set[str]]) -> List[str]:
+        return sorted(list(pins), key=int)
+
+
+# Add any new package types to both PACKAGE_DEFINITIONS and the PackageDef union
+PACKAGE_DEFINITIONS = {
+    "pga144": _PGAPackageDef(name="pga144", size=144)
+}
+PackageDef = Union[_PGAPackageDef, _BasePackageDef]
+
+class Port(pydantic.BaseModel):
+    type: str
+    pins: List[str]
+    direction: Optional[str] = None
+
+
+class Package(pydantic.BaseModel):
+    package_type: PackageDef = pydantic.Field(discriminator="type")
+    power: Dict[str, Port] = {}
+    clocks: Dict[str, Port] = {}
+    resets: Dict[str, Port] = {}
+
+    def check_pad(self, name: str, defn: dict):
+        match defn:
+            case {"type": "clock"}:
+                return self.clocks[name] if name in self.clocks else None
+            case {"type": "reset"}:
+                return self.resets[name] if name in self.clocks else None
+            case {"type": "source"}:
```

Comment:
although we do call them Vss and Vdd, we don't ever call them "source" or "drain", always "power" and "ground"

Reply:
Fair, I went with source and drain as 'power' is already the category and power.power felt weird ><

Reply:
i've left this as is for now, as we had some more converstions re power domains on element. would like to bottom those a bit before deciding the scheme here. make sense?

Reply:
Yeah, that does make sense, but I don't agree with sticking with this naming long term whatever we do end up going for

Reply:
agreed!

For now, I was thinking of changing to vdd/vss, how does that sound?